clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs

```diff
@@ -0,0 +1,135 @@
+            return Some(format!("{end_snippet} - {}", lower_bound - 1));
+        }
+        return Some(format!("{end_snippet} - {lower_bound}"));
+    }
+    None
+}
+
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    ex: &Expr<'_>,
+    receiver: &Expr<'_>,
+    arg: &Expr<'_>,
+    msrv: &Msrv,
+    method_call_span: Span,
+) {
+    let mut applicability = Applicability::MaybeIncorrect;
+    if let Some(range) = higher::Range::hir(receiver)
+        && let ExprKind::Closure(Closure { body, .. }) = arg.kind
+        && let body_hir = cx.tcx.hir().body(*body)
+        && let Body {
+            params: [param],
+            value: body_expr,
+        } = body_hir
+        && !usage::BindingUsageFinder::are_params_used(cx, body_hir)
+        && let Some(count) = extract_count_with_applicability(cx, range, &mut applicability)
+    {
+        let method_to_use_name;
+        let new_span;
+        let use_take;
+
+        if eager_or_lazy::switch_to_eager_eval(cx, body_expr) {
+            if msrv.meets(msrvs::REPEAT_N) {
+                method_to_use_name = "repeat_n";
+                let body_snippet = snippet_with_applicability(cx, body_expr.span, "..", &mut applicability);
+                new_span = (arg.span, format!("{body_snippet}, {count}"));
+                use_take = false;
+            } else {
+                method_to_use_name = "repeat";
+                let body_snippet = snippet_with_applicability(cx, body_expr.span, "..", &mut applicability);
+                new_span = (arg.span, body_snippet.to_string());
+                use_take = true;
+            }
+        } else if msrv.meets(msrvs::REPEAT_WITH) {
+            method_to_use_name = "repeat_with";
+            new_span = (param.span, String::new());
+            use_take = true;
+        } else {
+            return;
+        }
+
+        span_lint_and_then(
+            cx,
+            MAP_WITH_UNUSED_ARGUMENT_OVER_RANGES,
+            ex.span,
+            "map of a closure that does not depend on its parameter over a range",
+            |diag| {
+                diag.multipart_suggestion(
+                    if use_take {
+                        format!("remove the explicit range and use `{method_to_use_name}` and `take`")
+                    } else {
+                        format!("remove the explicit range and use `{method_to_use_name}`")
+                    },
+                    vec![
+                        (
+                            receiver.span.to(method_call_span),
+                            format!("std::iter::{method_to_use_name}"),
+                        ),
+                        new_span,
+                        (
+                            ex.span.shrink_to_hi(),
+                            if use_take {
+                                format!(".take({count})")
+                            } else {
+                                String::new()
+                            },
```

Comment:
These empty replacements (empty suggestion and empty span) trigger a [debug assertion in the compiler](https://github.com/rust-lang/rust/blob/e92993dbb43f0a5d17fe56e2d82f90435d6521c8/compiler/rustc_errors/src/diagnostic.rs#L922), so this needs to conditionally push the replacement into the vec

Reply:
I think it would be easier to just filter the vec. Let me know if you disagree stylistically.

Reply:
I don't think the filtering is easier than something like
```rs
let mut parts = vec![
    (receiver.span.to(method_call_span), format!("std::iter::{method_to_use_name}")),
    new_span,
];
if use_take {
    parts.push((ex.span.shrink_to_hi(), format!(".take({count})")));
}
```

I also find it a lot clearer because the current filtering makes me think that there could be other replacements that are empty when it's (as far as I can tell?) really only the last one, and the last one can only be empty because of an explicit `String::new()` so it would be IMO clearer to just not create an empty replacement like that in the first place and only push a non-empty replacement if necessary.

(The debug assertion also only exists to find bugs in suggestions, so it feels more like a workaround)